### ISSUE TITLE
compute vram consumption

### ISSUE THREAD

**Issue by goonm** [MIC-DKFZ/nnUNet#173]
Hi Fabian,
I was analyzing your code. But there was something difficult to understand.
By computing vram consumption, you decide network plan.

As a calculation method,
ref = Generic_UNet.use_this_for_batch_size_computation_3D 
here = Generic_UNet.compute_approx_vram_consumption(~~)
And compare "here" and "ref"(Recalculate until "ref" is bigger than "here").

But i think
ref_new = 12GB (case of 12GB GPU)
here_new = "here" * batch_size * 4byte(float32)
And compare "here_new" and "ref_new"(Recalculate until "ref_new" is bigger than "here_new").

What is your opinion?
Please explain the calculation process in more detail.

**Comment by FabianIsensee** [comment]
Hi,
the reference value is not an exact computation of GPU memory usage. It is proportional to GPU usage and that is good enough for me.
Best,
Fabian

**Comment by plesqui** [comment]
Hi Fabian, on a related note: 

If we were to run nnUNet on a GPU with more than 12GB of memory (say, a V100 with 32GB) and we would like to take advantage of the 32GB to determine the network parameters. Would just multiplying the value of `ref_new` by 32/12 be sufficient to achieve this goal? 

I guess I could also edit the value of batch size in the plans file to take advantage of the extra memory available. Have you had experience with any of these two approaches with the nnUnet?

Thank you for your repository! It is really great!

Cheers!
Pedro

**Comment by FabianIsensee** [comment]
Hi pedro,
both approaches will work. nnU-Net always prioritized patch size so by modifying ref_new you it will increase that first and foremost. Please allow it some headroom, like ref_new = ref_new / 12.5 * 32 or so (unless you feel like running some trial and error experiments ;-) )
I have experimented with both. For some datasets the results get better, for some worse (overfitting). Training times are annoyingly long. With the datasets that are currently available in out domain it will be hard to get large segmentation performance gains (due to their limited number of training cases). On other (internal) datasets I saw a larger performance gain
Best,
Fabian

**Comment by plesqui** [comment]
Hi Fabian,

Thank you so much for your feedback! 

all the best,

Pedro

**Comment by FabianIsensee** [comment]
Always happy to help
Best,
Fabian

**Comment by ahmedhosny** [comment]
Quick comment related to these numbers.

I have one 48MB GPU. I am adjusting the lines mentioned above, rerunning the experiment planning, and getting different batch sizes. It seems that larger batch sizes are causing longer epoch times. Is this the expected behavior? All of these were run on the same dataset and with the regular single GPU training (nnUNet_train V2) without DP or DDP.

batch size        | Memory usage           | multiplication factor  | time for first epoch 
------------- | ------------- | ---- | ------ 
2      |  ~7MB |  current implementation | ~400s 
 7     | ~18MB      |   24/Generic_UNet.BASE_NUM_FEATURES_3D | ~1300s
 14 | ~36MB  |     48/Generic_UNet.BASE_NUM_FEATURES_3D | ~2300s 
 18 | ~45MB |  64/Generic_UNet.BASE_NUM_FEATURES_3D | ~2800s 

I also noted that the lr does not change with batch size and wondering if it should be adjusted in the spirit of automated design? Thanks!

**Comment by ahmedhosny** [comment]
Now I see here https://github.com/MIC-DKFZ/nnUNet/issues/61 that self.num_batches_per_epoch is fixed - which explains this..

**Comment by FabianIsensee** [comment]
Hi,
I have not experimented with how to properly adjust number of batches and
and the arning rate which is why this is all fixed. As it is right now,
increasing the GPU budget should only be done if you expect segmentation
accuracy improvements as a result. If not then it is better to leave it at
default and enjoy the faster training :-)
You can of course tinker with the hyperparameters to (number of batches per
epoch. learning rate etc) but I am not sure whether you can speed up the
training by increasing the batch size on a single GPU.
nnU-Net always prioritizes patch size over batch size. The fact that
increasing the GPU budget increased the batch size only on your dataset
means that the patch size is already similar to the actual image size and
cannot be increased further.
It looks like you are training on a quado rtx 8000 (because of 48GB). You
will get much faster training by using a Volta GPU (Titan V, V100) because
Volta supports 3d conv acceleration with tensor cores. Due to a missing
kernel this is not supported on Turing (and I am constantly begging Nvidia
to implement this - without success so far ;-) )
Best,
Fabian
